core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -196,107 +268,141 @@ object H2OSchemaUtils {
+    val result = new ArrayBuffer[FieldWithOrder]()
+    var idx = 0
+    while (idx < values.length) {
+      val fieldQualifiedName = getQualifiedName(qualifiedName, idx.toString())
+      result ++= flattenField(fieldQualifiedName, elementType, nullable, metadata, values(idx), idx :: path)
+      idx = idx + 1
     }
+    result
   }
 
   private def flattenMapType(
+      qualifiedName: String,
       valueType: DataType,
-      containsNull: Boolean,
+      nullable: Boolean,
+      metadata: Metadata,
       data: Any,
-      path: List[Any],
-      qualifiedName: String,
-      nullableField: Boolean) = {
+      path: List[Any]) = {
     val map = data.asInstanceOf[Map[Any, Any]]
     val subRow = Row.fromSeq(map.values.toSeq)
-    map.keys.zipWithIndex.flatMap { case (key, idx) =>
-      val mapField = StructField(key.toString, valueType, containsNull)
-      flattenField(mapField, subRow(idx), key :: path, Some(qualifiedName), nullableField)
-    }.toSeq
+    val result = new ArrayBuffer[FieldWithOrder]()
+    map.foreach { case (key, value) =>
+      val fieldQualifiedName = getQualifiedName(qualifiedName, key.toString)
+      result ++= flattenField(fieldQualifiedName, valueType, nullable, metadata, value, key :: path)
+    }
+    result
   }
 
   private def flattenStructType(
+      qualifiedName: String,
+      nullableParent: Boolean,
+      metadata: Metadata,
       fields: Seq[StructField],
       data: Any,
-      path: List[Any],
-      qualifiedName: String,
-      nullableField: Boolean) = {
+      path: List[Any]) = {
     val subRow = data.asInstanceOf[Row]
     fields.zipWithIndex.flatMap { case (subField, idx) =>
-      flattenField(subField, subRow(idx), idx :: path, Some(qualifiedName), nullableField)
+      val StructField(name, dataType, nullable, fieldMetadata) = subField
+      val metadataBuilder = new MetadataBuilder()
+      metadataBuilder.withMetadata(metadata)
+      metadataBuilder.withMetadata(fieldMetadata)
+      val mergedMetedata = metadataBuilder.build()
+      val fieldQualifiedName = getQualifiedName(qualifiedName, name)
+      flattenField(fieldQualifiedName, dataType, nullable || nullableParent, mergedMetedata, subRow(idx), idx :: path)
     }
   }
 
-  def flattenStructsInSchema(schema: StructType, prefix: String = null, nullable: Boolean = false): StructType = {
+  def flattenStructsInSchema(
+      schema: StructType,
+      colPrefix: Option[String] = None,
+      newNamePrefix: Option[String] = None,
+      nullable: Boolean = false): Seq[(StructField, String)] = {
 
     val flattened = schema.fields.flatMap { f =>
       val escaped = if (f.name.contains(".")) "`" + f.name + "`" else f.name
-      val colName = if (prefix == null) escaped else prefix + "." + escaped
+      val colName = if (colPrefix.isDefined) colPrefix.get + "." + escaped else escaped
+      val newName = if (newNamePrefix.isDefined) newNamePrefix.get + "_" + f.name else f.name
```

Comment:
QQ: Why newName is introduced?

Reply:
The `newName` variable was introduced to enable `_` as a delimiter.

Reply:
If `_` is the new delimiter, do we need to keep `.` as delimiter?

Reply:
`.` is there to reference the original subcolumn in a `StructType`. This method just creates a mapping from a source to a destination.

Reply:
I see! What about renaming `colName` to `sourceColName` and `newName` to `targetNewName`? Or something like this?

Reply:
Yep, no problem :-)